caluma/caluma_form/structure.py

```diff
@@ -0,0 +1,218 @@
+                aq.question.slug: aq
+                for aq in self.children()
+                if aq.question.type == Question.TYPE_FORM
+            }
+        return self._sub_forms
+
+    def get_ans_question(
+        self, question_slug: str, check_parent: bool = True
+    ) -> Optional[Field]:
+
+        ans_question = self.ans_questions.get(question_slug)
+
+        if not ans_question and check_parent:
+            ans_question = (
+                self.parent().get_ans_question(question_slug) if self.parent() else None
+            )
+        if ans_question:
+            return ans_question
+
+        # OK start looking in subforms / row forms below our level.
+        # Since we're looking down, we're disallowing recursing to outer context
+        # to avoid recursing back to where we are
+        for subform in self.sub_forms.values():
+            sub_aq = subform.get_ans_question(question_slug, check_parent=False)
+            if sub_aq:
+                return sub_aq
+        # if we reach this line, we didn't find the question
+        return None
+
+    @object_local_memoise
+    def children(self):
+        answers = {ans.question_id: ans for ans in self.document.answers.all()}
+        return [
+            Field.factory(
+                document=self.document,
+                form=self.form,
+                question=question,
+                answer=answers.get(question.slug),
+                parent=self,
+            )
+            for question in self.form.questions.all()
+        ]
+
+    @object_local_memoise
+    def paths_to_question(self, slug):
+        res = []
+        prefix = [self.question] if self.question else []
+
+        if slug in self.ans_questions:
+            res.append(prefix + [self.ans_questions[slug].question])
+        for aq in self.children():
+
+            if aq.question.type == Question.TYPE_FORM:
+                for sub_path in aq.paths_to_question(slug):
+                    res.append(prefix + sub_path)
+
+            elif aq.question.type == Question.TYPE_TABLE:
+                for child in aq.children():
+                    for sub_path in child.paths_to_question(slug):
+                        # TODO: This should be covered, but is a case that will probably
+                        # never happen in real life
+                        res.append(prefix + sub_path)  # pragma: no cover
+        return res
+
+    def __repr__(self):
+        q_slug = self.question.slug if self.question else None
+        if q_slug:
+            return f"<FieldSet fq={q_slug}, doc={self.document.pk} hidden=({self.question.is_hidden}) req=({self.question.is_required})>"
+
+        return f"<FieldSet form={self.form.slug}, doc={self.document.pk}>"
+
+
+def print_document_structure(document):  # pragma: no cover
+    """Print a document's structure.
+
+    Intended halfway as an example on how to use the structure
+    classes, and a debugging utility.
+    """
+    ind = {"i": 0}
```

Comment:
Possibily stupid question, but I don't see it: Why is this an object instead of just an integer?

Reply:
Python scoping rules. Can't update variables in the "outer" scope by assignment. You could mess with the `global` keyword, but that's even worse

Reply:
Ah right, thanks!